### DISCUSSION TITLE
overlapping keybinds shouldn't count if they are below the current one

### DISCUSSION THREAD

**Discussion by Doksha** [ppy/osu#25639] [Ideas]
for example if you want to bind 7k mania from default to 
~~asd [space] hjk~~
dfk [space] kl;
you would be asked on s, d, j and k to confirm you want to change the key, which is very annoying (as you have to move your mouse to press and confirm, then click on the next key). ~~Lazer should only check for overlap above (before) the current key~~
Lazer should only prompt the user asking if they want to replace the key if it is above, otherwise assume they are about to rebind it and just set the conflicting key to none and let the user keep inputting keys.

**Comment by bdach** [comment]
The proposed solution makes negative sense. If the logic was written that way, you could assign multiple keybindings to one key just by binding them from the bottom up, defeating the purpose of the check.

Also your example doesn't make sense, since the defaults are <kbd>s</kbd><kbd>d</kbd><kbd>f</kbd><kbd>space</kbd><kbd>h</kbd><kbd>j</kbd><kbd>k</kbd>. Assigning all of the bindings you gave top to bottom in that order results in no binding conflicts.

**Reply by Doksha** [reply]
sorry I meant dfg [space] kl;

Just have it automatically overwrite (and set the other key to none) that way there are no conflicts

Is there really that much of a problem if you can bind one key to multiple actions, worst thing that that will do is make it possible to cheese really low star maps

**Reply by bdach** [reply]
> Just have it automatically overwrite (and set the other key to none) that way there are no conflicts

This was discussed and the conflict resolution popup was deemed the most user-friendly solution.

> Is there really that much of a problem if you can bind one key to multiple actions, worst thing that that will do is make it possible to cheese really low star maps

https://github.com/ppy/osu/issues/24780

**Reply by Doksha** [reply]
for singular options (ones that don't automatically move on to the next one) yes, in that case it would be more user friendly to ask, but in grouped keybinds that automatically progress (like mania) it can get very tedious to change keys (for example moving the keys to the right for 10k, especially with trackpad)

I don't see any reason these couldn't both work (have current system for singular options, and automatic overwriting for grouped options) 

Alternatively, next to the reset to defaults button, have a clear inputs button, that way you can keep the confirmation, but still make binding keys easy

**Reply by peppy** [reply]
Making the overwrite popover react to enter to overwrite may help? or arrow keys to decide which direction to overwrite. Something that doesn't require more clicks.

**Comment by itsMapleLeaf** [comment]
This is really irking me while I experiment with different similar key combinations to find something comfortable.

Suggestion: after assigning all of the bindings for section, if there are any remaining conflicts, _then_ show the bubble for each one, and, as suggested here, have it react to enter to accept, or escape to reject. For clarity, maybe show a `< x/2 >` on the dialog for when there are multiple. Bad hasty edit to show what I mean:
![](https://github.com/ppy/osu/assets/19603573/3b02df37-5896-4bce-b154-a8a88ac668d1)

**Reply by peppy** [reply]
Would it help if you can hit enter to confirm overwrite or something?

**Reply by itsMapleLeaf** [reply]
My ask (and I assume OP's ask as well) is to be able to fluidly set your keys for that section in one go, like one could before this dialog came in. Being able to hit enter would _help_, but I'd prefer a solution that keeps that capability.

**Reply by bdach** [reply]
The solution proposed above is not happening as it will get too complex. It is possible that resolving one conflict may spawn another one.

It's either enter or we just remove this popover and automatically resolve every binding conflict by unbinding the previous action from the newly-assigned key. Which seems like a UX regression to me buuuuut...

**Reply by itsMapleLeaf** [reply]
> we just remove this popover and automatically resolve every binding conflict by unbinding the previous action from the newly-assigned key

I would vote for this. I definitely get the desire for clarity, but I personally feel it causes too much friction to justify it

**Reply by peppy** [reply]
The ONLY way the dialog would be getting disabled is for ruleset bindings specifically. As in having it auto-unbind keys for within a ruleset makes sense to me but for any other global bindings it should remain as it fixes serious user-fuck-ups.

**Reply by itsMapleLeaf** [reply]
I'm fine with that!

**Comment by DaiterGG** [comment]
just to illustrate the problem, this is how I would change my 6k and 8k binds

https://github.com/ppy/osu/assets/64665908/462504a6-06a3-47fa-8a2b-3d1e8df40160

I don't propose to introduce any new behavior here, just copy stable.

**Comment by Hifumindesu** [comment]
One alternative solution is to add a second button that clears that entire section.
